internal/mesh/internal/controllers/sidecarproxy/cache/cache.go

```diff
@@ -0,0 +1,144 @@
+package cache
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/hashicorp/consul/internal/mesh/internal/types/intermediate"
+	"github.com/hashicorp/consul/internal/resource"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+)
+
+// Cache stores information needed for the sidecar-proxy controller to reconcile efficiently.
+// This currently means storing a list of all destinations for easy look up
+// as well as indices of source proxies where those destinations are referenced.
+//
+// It is the responsibility of the controller and its subcomponents (like mapper and data fetcher)
+// to keep this cache up-to-date as we're observing new data.
+type Cache struct {
+	lock sync.RWMutex
+
+	// store is a map from destination service reference and port as a string ID
+	// to the object representing destination reference.
+	store map[string]*intermediate.CombinedDestinationRef
+
+	// sourceProxiesIndex stores a map from a string representation of source proxy ID
+	// to the keys in the store map.
+	sourceProxiesIndex map[string]storeKeys
+}
+
+type storeKeys map[string]struct{}
+
+func New() *Cache {
+	return &Cache{
+		store:              make(map[string]*intermediate.CombinedDestinationRef),
+		sourceProxiesIndex: make(map[string]storeKeys),
+	}
+}
+
+func KeyFromID(id *pbresource.ID) string {
+	return fmt.Sprintf("%s/%s/%s",
+		resource.ToGVK(id.Type),
+		resource.TenancyToString(id.Tenancy),
+		id.Name)
+}
+
+func KeyFromRefAndPort(ref *pbresource.Reference, port string) string {
+	return fmt.Sprintf("%s:%s",
+		resource.ReferenceToString(ref),
```

Comment:
The comment I left on `ReferenceToString` implied the format had no perceived forward compatibility contract and was provided for logging/debug purposes only.

If you are going to rely upon this for an internal structural key either that comment should be amended, or you could consider changing to use struct keys here like:

```
type Key struct {
    resource.ReferenceKey
    Port string
}
```

and using `Key` instead of `string` in your maps.